# < Python程式設計基礎 >


C. Alex Hu @  2020/04/23

> - Python 是透過 interpreter 進行逐行直譯 (interpreting) 的程式語言，無須透過編譯 (compiling)。

> - Python 起源：由創始人 Guido van Rossum 於 1989 年起開發，名稱來自於英國 BBC show “Monty Python’s Flying Circus” (與 "蟒蛇 Python" 無關，參考 Python Official Site: https://docs.python.org/3/tutorial/appetite.html ) 。


## Chapter 1. 資料型別與基本運算
- [1.1 數值運算](#數值運算)
- [1.2 字串運算](#字串運算)

<a id='數值運算'></a>
### 1.1  數值運算
- 3 numeric types
    - integers 
    - floating point numbers
    - complex numbers
- Decimal
- Fraction

### Integers

In [1]:
2020    #  constant value of an integer

2020

In [2]:
2020 + 425         # 2445

2445

In [3]:
2020 - 4 * 25     # 1920

1920

### Floating Point Numbers

In [4]:
425.

425.0

In [5]:
2020 - 4 * 25.0  # coercion

1920.0

### Complex Numbers

In [6]:
3 + 5j

(3+5j)

### Numeric Operations

#### `int()` vs. `float()` : 資料型別轉換

In [7]:
int(2020 - 4 * 25.0)

1920

In [8]:
5 * 12.3

61.5

In [9]:
int(5 * 12.3)

61

In [10]:
float(2020)

2020.0

#### Division (/) always returns a floating point number.

In [11]:
(2020 - 4 * 25) / 5    # 384.0

384.0

In [12]:
100 / 5 

20.0

#### floor division //

In [13]:
100 // 3

33

#### modulus operator %

In [14]:
100 % 3  

1

####  ** operator to calculate powers 

In [15]:
1.732 ** 2

2.999824

In [16]:
10**5

100000

#### Assignment Operator , =
> To initiate the value and define the data type of a variable / an object.

> **`lValue`** **=** **`rValue`**
- **`rValue`** : Right Value (右值)，*value (值)*：10, 100.0, 'Hello world'
- **`lValue`** : Left Value (左值)，*address (址)*：x, myObject

In [17]:
interest_rate = 0.06
interest_rate

0.06

In [18]:
capital = 1000000
1 + interest_rate

1.06

#### The last printed expression is assigned to the variable _.

In [19]:
_ ** 32

6.4533866818853385

In [20]:
capital * _

6453386.681885338

### `Decimal` objects
- https://docs.python.org/3/library/decimal.html#decimal.Decimal

#### [NOTE]: There are some small differences between arithmetic on Decimal objects and arithmetic on integers and floats. 

In [21]:
(-15) % 4    ##  the remainder of (-15) % 4

1

In [22]:
(-15) // 4   ##  the quotiant of (-15) % 4

-4

- When the remainder operator % is applied to Decimal objects, the sign of the result is the sign of the dividend rather than the sign of the divisor:

In [23]:
from decimal import Decimal
Decimal(-15) % Decimal(4)

Decimal('-3')

- Similarly for // :

In [24]:
Decimal(-15) // Decimal(4)

Decimal('-3')

### `Fraction` objects
- https://docs.python.org/3/library/fractions.html#fractions.Fraction

In [25]:
from fractions import Fraction
Fraction(30, -4)

Fraction(-15, 2)

In [26]:
Fraction(2018)

Fraction(2018, 1)

In [27]:
Fraction()

Fraction(0, 1)

In [28]:
Fraction('4/23')

Fraction(4, 23)

In [29]:
Fraction('-4/23')

Fraction(-4, 23)

#### a finite floating number => a rational number = int / int

In [30]:
Fraction('1.41421')  

Fraction(141421, 100000)

In [31]:
Fraction('2.71828')

Fraction(67957, 25000)

In [32]:
Fraction('-.125')

Fraction(-1, 8)

In [33]:
Fraction('1.7e-3')

Fraction(17, 10000)

In [34]:
Fraction(3.33)

Fraction(1874623344892969, 562949953421312)

In [35]:
from decimal import Decimal
Fraction(Decimal('3.33'))    ##  Why?

Fraction(333, 100)

#### Why Fraction(333, 100) ? 
#### Ans : The Fraction class inherits from the abstract base class `numbers.Rational`, and implements all of the methods and operations from that class.

<a id='字串運算'></a>
### 1.2  字串運算

In [36]:
'Hello Python!'  # single quotes

'Hello Python!'

In [37]:
'isn\'t'  # use \' to escape the single quote...

"isn't"

In [38]:
"isn't"  # ...or use double quotes instead

"isn't"

In [39]:
'"Could you give me a ride, Alice?" Bob said.'

'"Could you give me a ride, Alice?" Bob said.'

In [40]:
"\"Sure. No problem.\" Alice answered."

'"Sure. No problem." Alice answered.'

In [41]:
'"No, it isn\'t," he said.'

'"No, it isn\'t," he said.'

#### Printing the string with `print()` function.

In [42]:
print('"No, it isn\'t," he said.')

"No, it isn't," he said.


In [43]:
string = 'Hello, Alice.\nHi, Bob. How are you?'  # \n means newline
string  

'Hello, Alice.\nHi, Bob. How are you?'

In [44]:
print(string)  # with print(), \n produces a new line

Hello, Alice.
Hi, Bob. How are you?


#### Use raw strings by adding an `r` before the first quote to avoid the escape character `\` :

In [45]:
print('C:\Python\notebook')  

C:\Python
otebook


In [46]:
print(r'C:\Python\notebook') 

C:\Python\notebook


#### Using triple-quotes: `"""..."""` or `'''...'''` to span string literals into multiple lines.

In [47]:
print("""\
[Agenda]
     9:00 ~  9:30               Reception
     9:30 ~ 10:20               Keynote Speech
""")

[Agenda]
     9:00 ~  9:30               Reception
     9:30 ~ 10:20               Keynote Speech



#### Strings' concatenation (+)  and repeatition (*):

In [48]:
'repea' + 'ti'*2 + 'on' 

'repeatition'

#### Two or more string literals next to each other are automatically concatenated.

In [49]:
'He' 'llo' ' Python' 

'Hello Python'

In [50]:
string = ("Fly Robin Fly" " Up Up To The Sky")
string

'Fly Robin Fly Up Up To The Sky'

In [51]:
string "Python"  # Error : can't concatenate a variable and a string literal

SyntaxError: invalid syntax (<ipython-input-51-32f11543abd3>, line 1)

In [52]:
'repea' ('ti'*2) 'on'     ##  'repea' + 'ti'*2 + 'on' 

SyntaxError: invalid syntax (<ipython-input-52-0d17f6c1facc>, line 1)

#### 字串長度 - len()

In [53]:
text = 'Hello Python'
len(text)

12

### `str` function

In [54]:
str(1000)   ##  Transform an integer into a string

'1000'

In [55]:
str(3.1415926) ##  Transform a floating point into a string

'3.1415926'